server/images.go

```diff
@@ -714,8 +716,7 @@ func deleteUnusedLayers(skipModelPath *ModelPath, deleteMap map[string]struct{})
 		// save (i.e. delete from the deleteMap) any files used in other manifests
 		manifest, _, err := GetManifest(fmp)
 		if err != nil {
-			//nolint:nilerr
-			return nil
```

Comment:
This was nil so a single bad manifest does not skip all other manifests

Reply:
If a manifest is unrecognized for any reason, the original version will cause all of the layer data associated with it to be deleted. That is sort of OK if it is pulled from the internet but not good as we allow users to create more of their own data. When it comes to data deletion I would prefer to be more conservative.

Reply:
We may want to return `filepath.SkipDir` here: https://pkg.go.dev/path/filepath#SkipDir so the walk can continue:

```
if the function returns a non-nil error, Walk stops entirely and returns that error.
```

source: https://pkg.go.dev/path/filepath#WalkFunc

Reply:
The goal of this section of the code is to remove layers that aren't referenced by any manifest file. If we want to be conservative about not deleting data that could be referenced by an unreadable manifest then we can't delete anything if we have one bad manifest since in theory it could reference anything. In this case, we just abort the whole delete operation and there isn't a benefit to continue walking through the manifests.

If we want to be more aggressive about cleaning up the disk then the original code is fine. This has the benefit of avoiding potentially filling up the disk with uncollected garbage data but personally I would err on the side of not deleting data.

In theory we could be more aggressive about deleting data that was pulled from the internet and could be refetched but that opens up a bunch of additional questions and I don't think is really worth it for an edge case.

Reply:
`SkipDir` doesn't help in this case since layers in a manifest are _removed_ from the delete map

Reply:
Ah right, sorry

Reply:
I think we can be aggressive here – if a manifest is malformed wouldn't it be safe to assume its layers are not going to be useful anyways and continue the cleanup?

Reply:
It depends on how valuable the data is - if it's sufficiently valuable then giving every chance possible for recovery is good. On a big enough scale, things happen. A bug could result in a corrupt manifest - would you rather be able to write a script to recover it or have all of the underlying data be gone?

To be honest this is hopefully a pretty fringe case so either choice doesn't have much impact. But in that unlikely scenario, if given the choice of a disk full error or deleted user data, I know which one I would choose.

One sort-of related example of auto-delete gone wild:
https://cloud.google.com/blog/products/infrastructure/details-of-google-cloud-gcve-incident

Reply:
I think it's ok here to err on the safe side and skip deleting if we are in a bad state.